pkg/plugin/validator.go

```diff
@@ -0,0 +1,171 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/abcxyz/pkg/githubapp"
+	"github.com/google/go-github/v55/github"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+)
+
+// Validator validates github issue against validation criteria.
+type Validator struct {
+	cfg        *PluginConfig
+	decodedPem *rsa.PrivateKey
+	client     *github.Client
+
+	// for testing only
+	fakeGithubIssue *github.Issue
+}
+
+// ExchangeResponse is the GitHub API response of requesting an access token
+// for the GitHub App installation with requested repositories and permissions.
+type ExchangeResponse struct {
+	AccessToken string `json:"token"`
+}
+
+// pluginGithubIssue contains the required attribute parsed from
+// the issue URL.
+type pluginGithubIssue struct {
+	Owner       string
+	RepoName    string
+	IssueNumber int
+}
+
+// NewValidator creates a validator.
+func NewValidator(cfg *PluginConfig) (*Validator, error) {
+	pk, err := readPrivateKey(cfg.GitHubAppPrivateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read private key: %w", err)
+	}
+	v := Validator{
+		cfg:        cfg,
+		decodedPem: pk,
+		client:     github.NewClient(nil),
+	}
+	return &v, nil
+}
+
+// MatchIssue parses issue info from provided issueURL and validate if the issue is valid.
+func (v *Validator) MatchIssue(ctx context.Context, issueURL string, opts ...githubapp.ConfigOption) error {
```

Comment:
Added a WithGIthubApp option, so githubApp can be directly provided.

Reply:
I am thinking instead of WithGIthubApp, maybe the new Validator can accept a githubApp as parameter, we can create the githubApp when we need to create the validator.

Reply:
The customized GitHubapp is only for testing purposes, I currently doesn't see it will be used elsewhere, so I think let the validator handle the githubapp makes more sense, as plugin.go can just call NewValidator(), and use its method.

Reply:
Update: discussed offline, decide to create githubApp and githubClient outside of validator for more centralized control over clients etc.